Gems/Archive/Code/Source/Clients/ArchiveReader.cpp

```diff
@@ -0,0 +1,928 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "ArchiveReader.h"
+
+#include <AzCore/IO/ByteContainerStream.h>
+#include <AzCore/IO/GenericStreams.h>
+#include <AzCore/IO/OpenMode.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+#include <AzCore/Task/TaskGraph.h>
+
+#include <Archive/ArchiveTypeIds.h>
+
+#include <Compression/DecompressionInterfaceAPI.h>
+
+namespace Archive
+{
+    // Implement TypeInfo, Rtti and Allocator support
+    AZ_TYPE_INFO_WITH_NAME_IMPL(ArchiveReader, "ArchiveReader", ArchiveReaderTypeId);
+    AZ_RTTI_NO_TYPE_INFO_IMPL(ArchiveReader, IArchiveReader);
+    AZ_CLASS_ALLOCATOR_IMPL(ArchiveReader, AZ::SystemAllocator);
+
+    ArchiveReader::ArchiveReader() = default;
+
+    ArchiveReader::~ArchiveReader()
+    {
+        UnmountArchive();
+    }
+
+    ArchiveReader::ArchiveReader(const ArchiveReaderSettings& writerSettings)
+        : m_settings(writerSettings)
+    {}
+
+    ArchiveReader::ArchiveReader(AZ::IO::PathView archivePath, const ArchiveReaderSettings& writerSettings)
+        : m_settings(writerSettings)
+    {
+        MountArchive(archivePath);
+    }
+
+    ArchiveReader::ArchiveReader(ArchiveStreamPtr archiveStream, const ArchiveReaderSettings& writerSettings)
+        : m_settings(writerSettings)
+    {
+        MountArchive(AZStd::move(archiveStream));
+    }
+
+    bool ArchiveReader::ReadArchiveHeader(ArchiveHeader& archiveHeader, AZ::IO::GenericStream& archiveStream)
+    {
+        // Read the Archive header into memory
+        AZStd::scoped_lock archiveLock(m_archiveStreamMutex);
+        archiveStream.Seek(0, AZ::IO::GenericStream::SeekMode::ST_SEEK_BEGIN);
+        AZ::IO::SizeType bytesRead = archiveStream.Read(sizeof(archiveHeader), &archiveHeader);
+        archiveStream.Seek(0, AZ::IO::GenericStream::SeekMode::ST_SEEK_BEGIN);
+        if (bytesRead != sizeof(archiveHeader))
+        {
+            m_settings.m_errorCallback({ ArchiveReaderErrorCode::ErrorReadingHeader, ArchiveReaderErrorString::format(
+                "Archive header should have size %zu, but only %llu bytes were read from the beginning of the archive",
+                sizeof(archiveHeader), bytesRead) });
+        }
+
+        return true;
+    }
+
+    // Define the Archive TableOfContentReader constructors
+    ArchiveReader::ArchiveTableOfContentsReader::ArchiveTableOfContentsReader() = default;
+
+    // Takes ownership of the buffer containing the Table of Contents raw data
+    // and an ArchiveTableOfContentsView instance
+    ArchiveReader::ArchiveTableOfContentsReader::ArchiveTableOfContentsReader(AZStd::vector<AZStd::byte> tocBuffer,
+        ArchiveTableOfContentsView tocView)
```

Comment:
Doesn't this need `&&` operators?
```suggestion
    ArchiveReader::ArchiveTableOfContentsReader::ArchiveTableOfContentsReader(AZStd::vector<AZStd::byte> && tocBuffer,
        ArchiveTableOfContentsView && tocView)
```

Reply:
No, the move happens internally using the value type.
If the caller wants to move the tocBuffer into the ArchiveTableOfContentsReader, the choice is left up to them with this API.
In their calling code `ArchiveTableOfContentsReader(AZStd::move(tocBuffer))`, then they would move the buffer contents
If they instead invoke `ArchiveTbleOfContentsReader(tocBuffer)`, they would make the copy of contents

If the signature was `AZStd::vector<AZStd::byte>&&` then the user would be forced to transfer ownership over which isn't explicitly required.

The comment above the function that says "Takes ownership of the buffer" is wrong though.

Reply:
yeah, I was basing this on the description

Reply:
I updated to the class constructor comment to remove the bit about taking ownership of the buffer.